File: session_cpp/src/vector_test.cpp
#include "catch_amalgamated.hpp"
#include "vector.h"
#include <cmath>
#include <filesystem>
#include <fstream>

namespace session_cpp {

TEST_CASE("Vector constructor.") {
  Vector v(1.0, 2.0, 3.0);
  REQUIRE(v.x() == 1.0);
  REQUIRE(v.y() == 2.0);
  REQUIRE(v.z() == 3.0);
  REQUIRE(!v.guid.empty());
}

TEST_CASE("Vector equality.") {
  Vector v1(1.0, 2.0, 3.0), v2(1.0, 2.0, 3.0), v3(1.1, 2.0, 3.0);
  REQUIRE(v1 == v2);
  REQUIRE_FALSE(v1 != v2);
  REQUIRE_FALSE(v1 == v3);
  REQUIRE(v1 != v3);
}

TEST_CASE("Vector to_json_data") {
  Vector v(15.5, 25.7, 35.9);
  v.name = "my_vector";
  auto data = v.to_json_data();
  REQUIRE(data["type"] == "Vector");
  REQUIRE(data["name"] == "my_vector");
  REQUIRE(std::abs(data["x"].get<float>() - 15.5f) < 1e-5f);
  REQUIRE(std::abs(data["y"].get<float>() - 25.7f) < 1e-5f);
  REQUIRE(std::abs(data["z"].get<float>() - 35.9f) < 1e-5f);
}

TEST_CASE("Vector from_json_data") {
  Vector orig(42.1, 84.2, 126.3);
  orig.name = "control_point_B";
  Vector rest = Vector::from_json_data(orig.to_json_data());
  REQUIRE(std::abs(rest.x() - 42.1f) < 1e-5f);
  REQUIRE(std::abs(rest.y() - 84.2f) < 1e-5f);
  REQUIRE(std::abs(rest.z() - 126.3f) < 1e-5f);
  REQUIRE(rest.name == "control_point_B");
  REQUIRE(rest.guid == orig.guid);
}

TEST_CASE("Vector to_json from_json") {
  Vector orig(123.45, 678.90, 999.11);
  orig.name = "file_test_point";
  orig.to_json("test_vector.json");
  Vector load = Vector::from_json("test_vector.json");
  REQUIRE(std::abs(load.x() - orig.x()) < 1e-5f);
  REQUIRE(std::abs(load.y() - orig.y()) < 1e-5f);
  REQUIRE(std::abs(load.z() - orig.z()) < 1e-5f);
  REQUIRE(load.name == orig.name);
  REQUIRE(load.guid == orig.guid);
}

TEST_CASE("Vector default constructor") {
  Vector v;
  REQUIRE(v[0] == 0);
  REQUIRE(v[1] == 0);
  REQUIRE(v[2] == 0);
}

TEST_CASE("Vector constructor") {
  Vector v(0.57, -158.63, 180.890);
  REQUIRE(std::abs(v[0] - 0.57f) < 1e-5f);
  REQUIRE(std::abs(v[1] + 158.63f) < 1e-5f);
  REQUIRE(std::abs(v[2] - 180.890f) < 1e-5f);
}

TEST_CASE("Vector static methods") {
  Vector x = Vector::x_axis(), y = Vector::y_axis(), z = Vector::z_axis();
  REQUIRE((x[0] == 1 && x[1] == 0 && x[2] == 0));
  REQUIRE((y[0] == 0 && y[1] == 1 && y[2] == 0));
  REQUIRE((z[0] == 0 && z[1] == 0 && z[2] == 1));
}

TEST_CASE("Vector from_start_and_end") {
  Vector v = Vector::from_start_and_end(Vector(8.7, 5.7, -1.87), Vector(1, 1.57, 2));
  REQUIRE(std::abs(v[0] + 7.7) < 1e-5f);
  REQUIRE(std::abs(v[1] + 4.13) < 1e-5f);
  REQUIRE(std::abs(v[2] - 3.87) < 1e-5f);
}

TEST_CASE("Vector operators and compound ops") {
  Vector v1(1, 2, 3);
  Vector v2(4, 5, 6);
  Vector v3 = v1 + v2;
  REQUIRE(v3[0] == 5);
  REQUIRE(v3[1] == 7);
  REQUIRE(v3[2] == 9);
  v3 = v1 - v2;
  REQUIRE(v3[0] == -3);
  REQUIRE(v3[1] == -3);
  REQUIRE(v3[2] == -3);
  v3 = v1 * 2;
  REQUIRE(v3[0] == 2);
  REQUIRE(v3[1] == 4);
  REQUIRE(v3[2] == 6);
  v3 = v1 / 2;
  REQUIRE(v3[0] == 0.5);
  REQUIRE(v3[1] == 1);
  REQUIRE(v3[2] == 1.5);
  v3 = v1;
  v3 += v2;
  REQUIRE(v3[0] == 5);
  REQUIRE(v3[1] == 7);
  REQUIRE(v3[2] == 9);
  v3 -= v2;
  REQUIRE(v3[0] == 1);
  REQUIRE(v3[1] == 2);
  REQUIRE(v3[2] == 3);
  v3 *= 2;
  REQUIRE(v3[0] == 2);
  REQUIRE(v3[1] == 4);
  REQUIRE(v3[2] == 6);
  v3 /= 2;
  REQUIRE(v3[0] == 1);
  REQUIRE(v3[1] == 2);
  REQUIRE(v3[2] == 3);
}

TEST_CASE("Vector reverse") {
  Vector v(1, 2, 3);
  v.reverse();
  REQUIRE((v[0] == -1 && v[1] == -2 && v[2] == -3));
}

TEST_CASE("Vector length") {
  Vector v(5.5697, -9.84, 1.587);
  float length = v.length();
  REQUIRE(std::round(length * 100) / 100 == 11.42f);
}

TEST_CASE("Vector unitize") {
  Vector v(5.5697, -9.84, 1.587);
  REQUIRE(v.unitized().length() == 1);
  v.unitize();
  REQUIRE(v.length() == 1);
}

TEST_CASE("Vector projection") {
  Vector v(1, 1, 1), x(1, 0, 0), y(0, 1, 0), z(0, 0, 1);
  auto [px, lenx, perp_x, plenx] = v.projection(x);
  auto [py, leny, perp_y, pleny] = v.projection(y);
  auto [pz, lenz, perp_z, plenz] = v.projection(z);
  REQUIRE((px[0] == 1 && px[1] == 0 && px[2] == 0));
  REQUIRE((py[0] == 0 && py[1] == 1 && py[2] == 0));
  REQUIRE((pz[0] == 0 && pz[1] == 0 && pz[2] == 1));
}

TEST_CASE("Vector is_parallel_to") {
  Vector v1(0, 0, 1), v2(0, 0, 2), v3(0, 0, -1), v4(0, 1, -1);
  REQUIRE(v1.is_parallel_to(v2) == 1);
  REQUIRE(v1.is_parallel_to(v3) == -1);
  REQUIRE(v1.is_parallel_to(v4) == 0);
}

TEST_CASE("Vector dot") {
  Vector v1(1, 0, 0), v2(0, 1, 0), v3(-1, 0, 0);
  REQUIRE(v1.dot(v2) == 0);
  REQUIRE(v1.dot(v3) == -1);
  REQUIRE(v1.dot(v1) == 1);
  
  float dot = v1.dot(v2), mag = v1.length() * v2.length();
  if (mag > 0.0f) {
    float angle_deg = std::acos(dot / mag) * geo::GLOBALS::TO_DEGREES;
    REQUIRE(std::round(angle_deg) == 90);
  }
}

TEST_CASE("Vector cross") {
  Vector v1(1, 0, 0), v2(0, 1, 0);
  Vector v3 = v1.cross(v2);
  REQUIRE((v3[0] == 0 && v3[1] == 0 && v3[2] == 1));
}

TEST_CASE("Vector angle") {
  Vector v1(1, 1, 0), v2(0, 1, 0);
  REQUIRE(std::abs(v1.angle(v2, false) - 45) < 1e-5f);
  REQUIRE(std::abs(Vector(-1, 1, 0).angle(v2, true) + 45) < 1e-5f);
}

TEST_CASE("Vector get_leveled_vector") {
  float scale = 1.0f;
  Vector lev = Vector(1, 1, 1).get_leveled_vector(scale);
  REQUIRE(std::round(lev.length() * 100) / 100 == 4.17f);
}

TEST_CASE("Vector cosine_law") {
  float a = 100, b = 150, angle = 115;
  float c = Vector::cosine_law(a, b, angle, true);
  REQUIRE(std::round(c * 100) / 100 == 212.55f);
}

TEST_CASE("Vector sine_law_angle") {
  float a = 212.55f, angle_a = 115, b = 150;
  float angle_b = Vector::sine_law_angle(a, angle_a, b);
  REQUIRE(std::round(angle_b * 100) / 100 == 39.76f);
}

TEST_CASE("Vector sine_law_length") {
  float a = 212.55f, angle_a = 115, angle_b = 39.761714f;
  float len_b = Vector::sine_law_length(a, angle_a, angle_b);
  REQUIRE(std::round(len_b * 100) / 100 == 150);
}

TEST_CASE("Vector angle_between_vector_xy_components") {
  Vector v1(std::sqrt(3), 1, 0), v2(1, std::sqrt(3), 0);
  REQUIRE(std::round(Vector::angle_between_vector_xy_components(v1) * 100) / 100 == 30);
  REQUIRE(std::round(Vector::angle_between_vector_xy_components(v2) * 100) / 100 == 60);
}

TEST_CASE("Vector sum_of_vectors") {
  std::vector<Vector> vecs = {Vector(1, 1, 1), Vector(2, 2, 2), Vector(3, 3, 3)};
  Vector sum = Vector::sum_of_vectors(vecs);
  REQUIRE((sum[0] == 6 && sum[1] == 6 && sum[2] == 6));
}

TEST_CASE("Vector coordinate_direction_angles") {
  auto abg = Vector(35.4, 35.4, 86.6).coordinate_direction_3angles(true);
  REQUIRE(std::abs(abg[0] - 69.274204) < 1e-4f);
  REQUIRE(std::abs(abg[1] - 69.274204) < 1e-4f);
  REQUIRE(std::abs(abg[2] - 30.032058) < 1e-4f);

  auto pt = Vector(1, 1, std::sqrt(2)).coordinate_direction_2angles(true);
  REQUIRE((std::abs(pt[0] - 45) < 1e-5f && std::abs(pt[1] - 45) < 1e-5f));
}

} // namespace session_cpp

File: session_rust/src/vector_test.rs
#[cfg(test)]
mod vector_tests {
    use crate::{globals, Vector};

    #[test]
    fn test_vector_constructor() {
        let v = Vector::new(1.0, 2.0, 3.0);
        assert_eq!((v.x(), v.y(), v.z()), (1.0, 2.0, 3.0));
        assert!(!v.guid.is_empty());
    }

    #[test]
    fn test_vector_equality() {
        let v1 = Vector::new(1.0, 2.0, 3.0);
        let mut v2 = Vector::new(1.0, 2.0, 3.0);
        v2.guid = v1.guid.clone();
        assert_eq!(v1, v2);
        let v3 = Vector::new(1.1, 2.0, 3.0);
        assert_ne!(v1, v3);
    }

    #[test]
    fn test_vector_to_json_data() {
        let v = Vector::new(15.5, 25.7, 35.9);
        let data = v.to_json_data().unwrap();
        assert!(data.contains("Vector") && data.contains("15.5"));
    }

    #[test]
    fn test_vector_from_json_data() {
        let orig = Vector::new(42.1, 84.2, 126.3);
        let rest = Vector::from_json_data(&orig.to_json_data().unwrap()).unwrap();
        assert_eq!((rest.x(), rest.y(), rest.z()), (42.1, 84.2, 126.3));
    }

    #[test]
    fn test_vector_to_json_from_json() {
        let orig = Vector::new(123.45, 678.90, 999.11);
        orig.to_json("test_vector.json").unwrap();
        let load = Vector::from_json("test_vector.json").unwrap();
        assert_eq!(
            (load.x(), load.y(), load.z()),
            (orig.x(), orig.y(), orig.z())
        );
    }

    #[test]
    fn test_vector_default_constructor() {
        let v = Vector::default();
        assert_eq!((v[0], v[1], v[2]), (0.0, 0.0, 0.0));
    }

    #[test]
    fn test_vector_constructor_values() {
        let v = Vector::new(0.57, -158.63, 180.890);
        assert_eq!((v[0], v[1], v[2]), (0.57, -158.63, 180.890));
    }

    #[test]
    fn test_vector_static_methods() {
        assert_eq!(
            (
                Vector::x_axis()[0],
                Vector::x_axis()[1],
                Vector::x_axis()[2]
            ),
            (1.0, 0.0, 0.0)
        );
        assert_eq!(
            (
                Vector::y_axis()[0],
                Vector::y_axis()[1],
                Vector::y_axis()[2]
            ),
            (0.0, 1.0, 0.0)
        );
        assert_eq!(
            (
                Vector::z_axis()[0],
                Vector::z_axis()[1],
                Vector::z_axis()[2]
            ),
            (0.0, 0.0, 1.0)
        );
    }

    #[test]
    fn test_vector_from_start_and_end() {
        let v =
            Vector::from_start_and_end(&Vector::new(8.7, 5.7, -1.87), &Vector::new(1.0, 1.57, 2.0));
        assert!((v[0] + 7.7).abs() < 1e-5);
        assert!((v[1] + 4.13).abs() < 1e-5);
        assert!((v[2] - 3.87).abs() < 1e-5);
    }

    #[test]
    fn test_vector_operators() {
        let v1 = Vector::new(1.0, 2.0, 3.0);
        let v2 = Vector::new(4.0, 5.0, 6.0);
        let v3 = &v1 + &v2;
        assert_eq!((v3[0], v3[1], v3[2]), (5.0, 7.0, 9.0));
        let v3 = &v1 - &v2;
        assert_eq!((v3[0], v3[1], v3[2]), (-3.0, -3.0, -3.0));
        let v3 = &v1 * 2.0;
        assert_eq!((v3[0], v3[1], v3[2]), (2.0, 4.0, 6.0));
        let v3 = &v1 / 2.0;
        assert_eq!((v3[0], v3[1], v3[2]), (0.5, 1.0, 1.5));

        let mut v3 = Vector::new(1.0, 2.0, 3.0);
        v3 += &v2;
        assert_eq!((v3[0], v3[1], v3[2]), (5.0, 7.0, 9.0));
        v3 -= &v2;
        assert_eq!((v3[0], v3[1], v3[2]), (1.0, 2.0, 3.0));
        v3 *= 2.0;
        assert_eq!((v3[0], v3[1], v3[2]), (2.0, 4.0, 6.0));
        v3 /= 2.0;
        assert_eq!((v3[0], v3[1], v3[2]), (1.0, 2.0, 3.0));
    }

    #[test]
    fn test_vector_reverse() {
        let mut v = Vector::new(1.0, 2.0, 3.0);
        v.reverse();
        assert_eq!((v[0], v[1], v[2]), (-1.0, -2.0, -3.0));
    }

    #[test]
    fn test_vector_length() {
        let v = Vector::new(5.5697, -9.84, 1.587);
        let length = v.compute_length();
        assert_eq!(length, 11.417_781);
    }

    #[test]
    fn test_vector_unitize() {
        let v = Vector::new(5.5697, -9.84, 1.587);
        assert_eq!(v.normalize().compute_length(), 1.0);
        let mut v = Vector::new(5.5697, -9.84, 1.587);
        v.normalize_self();
        assert_eq!(v.magnitude(), 1.0);
    }

    #[test]
    fn test_vector_projection() {
        let v = Vector::new(1.0, 1.0, 1.0);
        let x = Vector::x_axis();
        let y = Vector::y_axis();
        let z = Vector::z_axis();
        let (proj_x, _lenx, _perp_x, _perp_lenx) = v.projection(&x);
        let (proj_y, _leny, _perp_y, _perp_leny) = v.projection(&y);
        let (proj_z, _lenz, _perp_z, _perp_lenz) = v.projection(&z);
        assert_eq!((proj_x[0], proj_x[1], proj_x[2]), (1.0, 0.0, 0.0));
        assert_eq!((proj_y[0], proj_y[1], proj_y[2]), (0.0, 1.0, 0.0));
        assert_eq!((proj_z[0], proj_z[1], proj_z[2]), (0.0, 0.0, 1.0));
    }

    #[test]
    fn test_vector_is_parallel_to() {
        let v1 = Vector::new(0.0, 0.0, 1.0);
        let v2 = Vector::new(0.0, 0.0, 2.0);
        let v3 = Vector::new(0.0, 0.0, -1.0);
        let v4 = Vector::new(0.0, 1.0, -1.0);
        assert_eq!(v1.is_parallel_to(&v2), 1);
        assert_eq!(v1.is_parallel_to(&v3), -1);
        assert_eq!(v1.is_parallel_to(&v4), 0);
    }

    #[test]
    fn test_vector_dot() {
        let v1 = Vector::new(1.0, 0.0, 0.0);
        let v2 = Vector::new(0.0, 1.0, 0.0);
        let v3 = Vector::new(-1.0, 0.0, 0.0);
        assert_eq!(v1.dot(&v2), 0.0);
        assert_eq!(v1.dot(&v3), -1.0);
        assert_eq!(v1.dot(&v1), 1.0);

        let dot = v1.dot(&v2);
        let mag = v1.compute_length() * v2.compute_length();
        if mag > 0.0 {
            let angle_deg = (dot / mag).acos() * unsafe { globals::TO_DEGREES as f32 };
            assert_eq!(angle_deg, 90.0);
        }
    }

    #[test]
    fn test_vector_cross() {
        let v1 = Vector::new(1.0, 0.0, 0.0);
        let v2 = Vector::new(0.0, 1.0, 0.0);
        let v3 = v1.cross(&v2);
        assert_eq!((v3[0], v3[1], v3[2]), (0.0, 0.0, 1.0));
    }

    #[test]
    fn test_vector_angle() {
        let v1 = Vector::new(1.0, 1.0, 0.0);
        let v2 = Vector::new(0.0, 1.0, 0.0);
        unsafe {
            assert!((v1.angle(&v2, false) - 45.0).abs() < globals::ZERO_TOLERANCE as f32);
            assert!(
                (Vector::new(-1.0, 1.0, 0.0).angle(&v2, true) + 45.0).abs()
                    < globals::ZERO_TOLERANCE as f32
            );
        }
    }

    #[test]
    fn test_vector_get_leveled_vector() {
        let lev = Vector::new(1.0, 1.0, 1.0).get_leveled_vector(1.0);
        assert!((lev.compute_length() - 4.168_432_5).abs() < 1e-4);
    }

    #[test]
    fn test_vector_cosine_law() {
        let c = Vector::cosine_law(100.0, 150.0, 115.0, true);
        assert_eq!((c * 100.0).round() / 100.0, 212.55);
    }

    #[test]
    fn test_vector_sine_law_angle() {
        let angle_b = Vector::sine_law_angle(212.55, 115.0, 150.0, true);
        assert_eq!((angle_b * 100.0).round() / 100.0, 39.76);
    }

    #[test]
    fn test_vector_sine_law_length() {
        let len_b = Vector::sine_law_length(212.55, 115.0, 39.761714, true);
        assert_eq!((len_b * 100.0).round() / 100.0, 150.0);
    }

    #[test]
    fn test_vector_angle_between_vector_xy_components() {
        let v1 = Vector::new(3.0_f32.sqrt(), 1.0, 0.0);
        let v2 = Vector::new(1.0, 3.0_f32.sqrt(), 0.0);
        assert_eq!(
            (Vector::angle_between_vector_xy_components(&v1) * 100.0).round() / 100.0,
            30.0
        );
        assert_eq!(
            (Vector::angle_between_vector_xy_components(&v2) * 100.0).round() / 100.0,
            60.0
        );
    }

    #[test]
    fn test_vector_sum_of_vectors() {
        let vecs = vec![
            Vector::new(1.0, 1.0, 1.0),
            Vector::new(2.0, 2.0, 2.0),
            Vector::new(3.0, 3.0, 3.0),
        ];
        let sum = Vector::sum_of_vectors(&vecs);
        assert_eq!((sum[0], sum[1], sum[2]), (6.0, 6.0, 6.0));
    }

    #[test]
    fn test_vector_coordinate_direction_angles() {
        let abg = Vector::new(35.4, 35.4, 86.6).coordinate_direction_3angles(true);
        assert!((abg[0] - 69.274_2).abs() < 1e-4);
        assert!((abg[1] - 69.274_2).abs() < 1e-4);
        assert!((abg[2] - 30.032058).abs() < 1e-4);

        let pt = Vector::new(1.0, 1.0, 2.0_f32.sqrt()).coordinate_direction_2angles(true);
        assert!((pt[0] - 45.0).abs() < 1e-6);
        assert!((pt[1] - 45.0).abs() < 1e-6);
    }
}
